src/main/scala/SimDRAM.scala

```diff
@@ -0,0 +1,44 @@
+package testchipip
+
+import chisel3._
+import chisel3.experimental.IntParam
+import chisel3.util.HasBlackBoxResource
+import freechips.rocketchip.amba.axi4.{AXI4BundleParameters, AXI4Bundle}
+import freechips.rocketchip.subsystem.{ExtMem, CacheBlockBytes, CanHaveMasterAXI4MemPortModuleImp}
+
+class SimDRAM(memSize: BigInt, lineSize: Int,
+              params: AXI4BundleParameters) extends BlackBox(Map(
+    "MEM_SIZE" -> IntParam(memSize),
+    "LINE_SIZE" -> IntParam(lineSize),
+    "ADDR_BITS" -> IntParam(params.addrBits),
+    "DATA_BITS" -> IntParam(params.dataBits),
+    "ID_BITS" -> IntParam(params.idBits))) with HasBlackBoxResource {
+
+  val io = IO(new Bundle {
+    val clock = Input(Clock())
+    val reset = Input(Bool())
+    val axi = Flipped(new AXI4Bundle(params))
+  })
+
+  require(params.dataBits <= 64)
+
+  addResource("/testchipip/vsrc/SimDRAM.v")
+  addResource("/testchipip/csrc/SimDRAM.cc")
+  addResource("/testchipip/csrc/mm.cc")
+  addResource("/testchipip/csrc/mm.h")
+}
+
+trait CanHaveBlackBoxSimMemModuleImp { this: CanHaveMasterAXI4MemPortModuleImp =>
```

Comment:
Is it necessary for this to be a trait? Why can't we collapse this entirely within an IOBinder?
```
class WithBlackBoxSimMem extends OverrideIOBinder({
  (c, r, s, top: CanHaveMasterAXI4MemPortModuleImp) => {
    val p = top.p
    (top.mem_axi4 zip top.outer.memAXI4Node).foreach { case (io, node) =>
      val memSize = p(ExtMem).get.master.size
      val lineSize = p(CacheBlockBytes)
      (io zip node.in).foreach { case (axi4, (_, edge)) =>
        val mem = Module(new SimDRAM(memSize, lineSize, edge.bundle))
        mem.io.axi <> axi4
        mem.io.clock := clock
        mem.io.reset := reset
      }
    }
  }
})
```

Reply:
Oh, I guess that's true.